### PR TITLE
Add a Slots library, and a transient variant of ReentrancyGuard

### LINKED ISSUE #4205: `ReentrancyGuard` with transient opcodes `TSTORE` and `TLOAD`
**🧐 Motivation**

Since [EIP-1153](https://eips.ethereum.org/EIPS/eip-1153) is confirmed to be included in the upcoming Cancun-Deneb upgrade, I would like to discuss the possibility of using the future transient opcodes `TSTORE` and `TLOAD` for the `ReentrancyGuard` (abstract) contract.

Furthermore, I would like to highlight the initiatives (partially driven by myself), to disable reentrancy by default at the compiler level:
- **Vyper:** https://github.com/vyperlang/vyper/issues/3380
- **Solidity:** https://github.com/ethereum/solidity/issues/12996

### LINKED ISSUE #4888: Use transient storage in ReentrancyGuard
For backward compatibility, and support of chains that do no implement EIP-1153, we whould have two version (the current one + a new transient one). Current one would be marked as deprecated

### PR DESCRIPTION
**Pull request by Amxx**
Fixes #4205
Fixes #4888

Note:
- **Hardhat update to 2.21.0:** Technically only 2.20.0 is needed to support cancun, but 2.21.0 includes EDR that improve test speed
- **hardhat.config.js:** Add a new entry `evm` / `evmVersion` that affects both the compiler (targeted evm version) and the node (executed evm version)
- **foundry.toml:** enable cancun (and set compiler to 0.8.24)

#### PR Checklist

- [x] Tests
- [ ] Documentation
- [ ] Changeset entry (run `npx changeset add`)

### CODE REVIEW

File: contracts/utils/TranscientReentrancyGuard.sol [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import {TranscientStorage} from "./TranscientStorage.sol";
+
+/**
+ * @dev Variant of {ReentrancyGuard} that uses transcient storage.
+ *
+ * NOTE: This variant only works on networks where EIP-1153 is available.
+ */
+abstract contract TranscientReentrancyGuard {
+    // keccak256(abi.encode(uint256(keccak256("openzeppelin.storage.ReentrancyGuard")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant ReentrancyGuardStorageLocation =
```
Comment (Amxx) [inline contracts/utils/TranscientReentrancyGuard.sol:14]:
```suggestion
    bytes32 private constant REENTRANCY_GUARD_STORAGE =
```

File: scripts/generate/templates/TranscientStorage.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
+const format = require('../format-lines');
+const { capitalize } = require('../../helpers');
+
+const TYPES = ['address', 'bool', 'bytes32', 'uint256'];
+
+const header = `\
+pragma solidity ^0.8.24;
+
+/**
+ * @dev Library for reading and setting transcient storage.
```
Comment (Amxx) [inline scripts/generate/templates/TranscientStorage.js:10]:
```suggestion
 * @dev Library for reading and writting transcient storage.
```

**Review by allwin199 — COMMENTED** [review]
@Amxx There is a typo in `writting` it should be `writing`

File: scripts/generate/templates/TransientStorage.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  /// @solidity memory-safe-assembly
+  assembly {
+      r := tload(slot)
+  }
+}
+`;
+
+const store = type => `\
+/**
+ * @dev Store \`value\` in the transient slot \`store\`.
```
Comment (ernestognw) [inline scripts/generate/templates/TransientStorage.js:30]:
```suggestion
 * @dev Store a \`${type}`\ value in the transient slot \`store\`.
```

File: contracts/utils/TransientReentrancyGuard.sol [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+pragma solidity ^0.8.24;
+
+import {TransientStorage} from "./TransientStorage.sol";
+
+/**
+ * @dev Variant of {ReentrancyGuard} that uses transient storage.
+ *
+ * NOTE: This variant only works on networks where EIP-1153 is available.
+ */
+abstract contract TransientReentrancyGuard {
```
Comment (ernestognw) [inline contracts/utils/TransientReentrancyGuard.sol:12]: What about `ReentrancyGuardTransient` instead? I feel the "transient" prefix captures the show:

```suggestion
abstract contract ReentrancyGuardTransient {
```

File: scripts/generate/templates/TransientStorage.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * Note: This library only works on networks where EIP-1153 is available.
+ */
+`;
+
+const load = type => `\
+/**
+ * @dev Returns the transient slot \`slot\` as a \`${type}\`.
+ */
+function load${capitalize(type)}(bytes32 slot) internal view returns (${type} r) {
```
Comment (ernestognw) [inline scripts/generate/templates/TransientStorage.js:20]: Any reason in particular to make the return value different to `StorageSlot`? The values there are defined as:

```
 struct Uint256Slot {
    uint256 value;
}
```

I definitely think this syntax is better, but I'm wondering if we should prefer one syntax over the other.

File: scripts/generate/templates/TransientStorage.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      r := tload(slot)
+  }
+}
+`;
+
+const store = type => `\
+/**
+ * @dev Store \`value\` in the transient slot \`store\`.
+ */
+function store(bytes32 slot, ${type} value) internal {
```
[the remark this answers is not in the document]
Comment (Amxx) [reply scripts/generate/templates/TransientStorage.js:32]: Why? We don't need that.

[the remark this answers is not in the document]
Comment (ernestognw) [reply scripts/generate/templates/TransientStorage.js:32]: Right. I was originally going to propose `tstore<Type>` as the name, and only `tstore` would conflict with the Yul instruction.

Let's solve this, I'm still making myself an opinion about its usage

File: scripts/generate/templates/TransientStorage.js (hunk 5 above)
Comment (Amxx) [reply scripts/generate/templates/TransientStorage.js:20]: AFAIK, Solidity supports structs in memory and in storage, but not transient ones. You can't mark it. Only thing we have is low level opcode access through assembly ?

File: scripts/generate/templates/TransientStorage.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Amxx) [reply scripts/generate/templates/TransientStorage.js:30]:
```suggestion
 * @dev Store ${type} \`value\` in the transient \`slot\` store.
```

File: scripts/generate/templates/TransientStorage.js (hunk 5 above)
Comment (Amxx) [reply scripts/generate/templates/TransientStorage.js:20]: Capture d’écran du 2024-03-14 20-49-50

Comment (Amxx) [reply scripts/generate/templates/TransientStorage.js:20]: Capture d’écran du 2024-03-14 20-50-54

File: contracts/utils/TransientStorage.sol
Comment (frangio) [inline contracts/utils/TransientStorage.sol:1]: Wouldn't it be nicer to provide an API more similar to the `StorageSlot` library?

A family of UDVT that encapsulate a transient location with `load()` and `store()` methods.

Comment (Amxx) [reply contracts/utils/TransientStorage.sol:1]: Should we merge the two? (put the transient related tools in `StorageSlot` ?)

Comment (Amxx) [reply contracts/utils/TransientStorage.sol:1]: I'm starting to think of something like that: https://gist.github.com/Amxx/86fc66c375a0ca73fa6d851066084466

File: scripts/generate/templates/TransientStorage.js (hunk 5 above)
Comment (Amxx) [reply scripts/generate/templates/TransientStorage.js:20]: See https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4955/files#r1525423381

File: .changeset/honest-planets-sell.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+---
+'openzeppelin-solidity': minor
+---
+
+`TransientStorage`: Added a library to help read and write to transient storage (EIP-1153).
```
Comment (Amxx) [inline .changeset/honest-planets-sell.md:5]:
```suggestion
`Slots`: Added a library to help derivate, read and write to storage slots (both "normal" and transient).
```

File: .changeset/hungry-oranges-travel.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+'openzeppelin-solidity': minor
+---
+
+`TransientReentrancyGuard`: Added a variant of `ReentrancyGuard` that uses transient storage.
```
Comment (Amxx) [inline .changeset/hungry-oranges-travel.md:5]:
```suggestion
`ReentrancyGuardTransient`: Added a variant of `ReentrancyGuard` that uses transient storage.
```

File: contracts/utils/README.adoc [hunk 9]
```diff
@@ -13,6 +13,7 @@ Miscellaneous contracts and libraries containing utility functions you can use t
  * {MerkleProof}: Functions for verifying https://en.wikipedia.org/wiki/Merkle_tree[Merkle Tree] proofs.
  * {EIP712}: Contract with functions to allow processing signed typed structure data according to https://eips.ethereum.org/EIPS/eip-712[EIP-712].
  * {ReentrancyGuard}: A modifier that can prevent reentrancy during certain functions.
+ * {TransientReentrancyGuard}: Variant of {ReentrancyGuard} that uses transient storage (EIP-1153).
```
Comment (Amxx) [inline contracts/utils/README.adoc:16]:
```suggestion
 * {ReentrancyGuardTransient}: Variant of {ReentrancyGuard} that uses transient storage (EIP-1153).
```

File: contracts/utils/README.adoc [hunk 10]
```diff
@@ -29,6 +30,7 @@ Miscellaneous contracts and libraries containing utility functions you can use t
  * {Strings}: Common operations for strings formatting.
  * {ShortString}: Library to encode (and decode) short strings into (or from) a single bytes32 slot for optimizing costs. Short strings are limited to 31 characters.
  * {StorageSlot}: Methods for accessing specific storage slots formatted as common primitive types.
+ * {TransientStorage}: Methods for reading and writing to transient storage.
```
Comment (Amxx) [inline contracts/utils/README.adoc:33]:
```suggestion
 * {Slots}: Methods for derivating, reading and writing to storage slots. This supports both "normal" and transient storage.
```

File: contracts/utils/README.adoc [hunk 11]
```diff
@@ -64,6 +66,8 @@ Because Solidity does not support generic types, {EnumerableMap} and {Enumerable
 
 {{ReentrancyGuard}}
 
+{{TransientReentrancyGuard}}
```
Comment (Amxx) [inline contracts/utils/README.adoc:69]:
```suggestion
{{ReentrancyGuardTransient}}
```

File: contracts/utils/README.adoc [hunk 12]
```diff
@@ -110,6 +114,8 @@ Ethereum contracts have no native concept of an interface, so applications must
 
 {{StorageSlot}}
 
+{{TransientStorage}}
```
Comment (Amxx) [inline contracts/utils/README.adoc:117]:
```suggestion
{{Slots}}
```

File: contracts/utils/Slots.sol
Comment (Amxx) [inline contracts/utils/Slots.sol:1]: @ernestognw @frangio Like this new approach?

**Comment by Amxx** [comment]
Note: `StorageSlot` is more powerfull that the proposed `Slot`, because it supports operations sur as `+=`, `|=` or `&=`. It also supports non value types (because solidity is in charge of the code generation for reading/writting to the "slot", we just force its location).

A few question
- I think the slot derivation function are super usefull, but I'm wondering where to put them. Maybe they deserve a dedicated file.
- I think adding sstore and sload doesn't add complexity (just file length), so I'd keep that for completeness, even though it somehow overlaps with `StorageSlot`.
- Maybe `Slot.sol` needs a different name. Ideas?

**Comment by frangio** [comment]
>I think adding sstore and sload doesn't add complexity (just file length), so I'd keep that for completeness, even though it somehow overlaps with StorageSlot.

I actually think it's better to remove those methods, and make the library purely about transient storage. It makes it a more sound abstraction for transient pointers, where there is only one correct dereferencing operation, and where it is not correct to cast into a normal storage pointer.

If those are removed `TransientSlot` may be a better name for the library. I think it's probably okay to leave the derivation methods in there since they are not necessary for storage slots? I'm not opposed to putting them in a separate file, but it would make it a little less convenient to use, since the current library only requires `using TransientSlot for *`.

**Comment by Amxx** [comment]
THe derivation function have a use for normal storage, see the diff to `contracts/utils/Arrays.sol`

**Pull request closed without merging**